### PR TITLE
riscv: fix-up non `.got.zig` symbol relocations

### LINKED ISSUE #20871: `__init_array_start` / `__init_array_end` logic disabled for riscv backend
https://github.com/ziglang/zig/blob/f219286573a7a1edafe3e1b5bc1e521a379ee2e2/lib/std/start.zig#L468-L469

Current status:

```
$ stage3/bin/zig test ../test/behavior.zig -target riscv64-linux -fno-llvm -fno-lld -mcpu=baseline+v+zbb
/home/andy/src/zig/lib/std/start.zig:376:1: error: TODO: div_exact
fn posixCallMainAndExit(argc_argv_ptr: [*]usize) callconv(.C) noreturn {
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by Rexicon226**
closes #20871

### CODE REVIEW

File: src/arch/riscv64/Emit.zig [hunk 1]
```diff
@@ -63,6 +63,9 @@ pub fn emitMir(emit: *Emit) Error!void {
 
                         hi_r_type = Elf.R_ZIG_GOT_HI20;
                         lo_r_type = Elf.R_ZIG_GOT_LO12;
+                    } else if (sym.flags.needs_got) {
+                        hi_r_type = Elf.R_GOT_HI20_STATIC; // TODO this needs reworking here and for x86_64
```
Comment (andrewrk) [inline src/arch/riscv64/Emit.zig:67]: "TODO this is needs to be rewritten" is a conclusion with a missing argument. Code comments should stick to providing facts and explanations, and avoid vague wishful thinking. Discussion about whether something should be rewritten or not belongs on the issue tracker, not in TODO comments.

#363

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: src/arch/riscv64/Emit.zig (hunk 1 above)
Comment (Rexicon226) [reply src/arch/riscv64/Emit.zig:67]: Fair point. Will open an issue and link to it.

File: src/link/Elf.zig [hunk 2]
```diff
@@ -6059,11 +6059,40 @@ const RelaSection = struct {
...
+    };
+
+    const other_relocs = .{
+        elf.R_X86_64,
+        elf.R_AARCH64,
+        elf.R_RISCV,
+        elf.R_PPC64,
+    };
+
+    @setEvalBranchQuota(other_relocs.len * zig_relocs.len * 256);
```
Comment (andrewrk) [inline src/link/Elf.zig:6087]: please cap this with an integer literal so we can see how much work it's doing

File: src/arch/riscv64/Emit.zig (hunk 1 above)
Comment (kubkon) [reply src/arch/riscv64/Emit.zig:67]: That's mainly a TODO for me and added by me. Re-engineering a solution at this stage is difficult as requires redesigning the concept of `.got.zig`.

**Review by andrewrk — APPROVED** [review]

File: src/arch/riscv64/Emit.zig (hunk 1 above)
Comment (andrewrk) [reply src/arch/riscv64/Emit.zig]: That's fine to delay re-engineering a solution. However, the comment should either explain the problem, link to an open issue, or be deleted.

**Pull request merged**